content/docs/reference/design-docs/editoast-errors/_index.en.md

```diff
@@ -0,0 +1,320 @@
+
+## Full `derive(EditoastError)` spec
+
+The macro supports enums, named structs and tuple structs (fixme correct naming).
+
+````rust
+#[derive(thiserror::Error)] // not an EditoastError
+#[error("wrong string: {0}")]
+struct WrongString(String);
+
+#[derive(EditoastError, thiserror::Error)]
+#[error("wrong int: {value}")]
+#[editoast_error(user, name = "InvalidInt")]
+struct WrongInt { value: i64 };
+
+#[derive(EditoastError, thiserror::Error)]
+#[error("my error type")]
+#[editoast_error(context)]
+enum MyError {
+    // error_type = "MyError::InvalidString"
+    // context = { expected_format: string }
+    // #[editoast_error(internal)] by default
+    InvalidString { #[from] source: WrongString, expected_format: String },
+
+    // error_type = "MyError::InvalidInt"
+    // context = {} (xyz is skipped as we forward the editoast error)
+    WrongInt { #[from] #[editoast_error] source: WrongInt, xyz: String }
+
+    // error_type = "MyError::Bad"
+    // context = { "0": string, "1": number }
+    #[error("user did a bad with {0} and {1}")]
+    #[editoast_error(user, name = "Bad")]
+    Oops(String, i64)
+}
+````
+
+### Providing `context`
+
+Only `4xx` errors require context since these are the errors the frontend may use to provide the user some form of recovery. Context is computed just before the error is serialized in `axum`'s error handler.
+
+`derive(EditoastError)` provides a few ways to set it.
+
+````rust
+#[derive(Debug, thiserror::Error, EditoastError)]
+enum Error {
+    #[editoast_error(user, context)]
+    No { because: String }, // { }
+
+    #[editoast_error(user, context)]
+    Nein { reasons: Vec<String> }, // { "reasons": [string] }
+
+    // explicitely builds the dictionnary (*)
+    #[editoast_error(user, context(
+        reason = "format!(\"because {reason}\")",
+        recovery = "format!(\"do this instead {recovery}\")"
+    ))]
+    QueNo { reason: String, recovery: String },
+}
+
+// with a provider function
+#[derive(Debug, thiserror::Error, EditoastError)]
+#[editoast_error(context_with = context_provider)]
+enum Error {
+    Nao(String),
+    Nee(String, u64)
+}
+
+fn context_provider(error: Error) -> HashMap<String, serde_json::Value> {
+    todo!()
+}
+````
+
+`(*)`: only if the mapping can be collected easily using `darling`.
+
+### Incident reports
+
+For `internal` errors that won't contain meaningful information for the end user, we substitute the error by:
+
+```json
+{
+  "error_type": "InternalError",
+  "message": "<a meaningful message>",
+  "status": 500,
+  "context": {},
+  "incident": "<uuid>"
+}
+```
+
+In order to be able to find and investigate the error later on, we associate to each `5xx` error a unique `incident` identifier. At first we'll just log the incident with:
```

Comment:
It's unclear what the use for this unique identifier. I would understand to have one if I could relate to other places where this identifier is shown. If the idea is to have that identifier so we can go to Datadog to find the entire related trace, then maybe this identifier can be the trace ID directly?

Reply:
> this identifier can be the trace ID directly

Do we have traceids when tracing isn't enabled?

Reply:
After the discussion we had during the PIP, I moved this in the "Rejected ideas" section, I'll let you resolve if its ok [`717e31d` (#237)](https://github.com/OpenRailAssociation/osrd-website/pull/237/commits/717e31df20a3a87ae897ed7212f155b6fa3d5e72)